File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/resolution/SemanticRequestResolver.java
package com.foundgine.core.semantic.resolution;

import com.foundgine.core.abstractions.*;
import com.foundgine.core.semantic.*;
import com.foundgine.core.semantic.query.*;
import java.util.*;

/**
 * Resolves a protocol-neutral SemanticRequest against the static semantic model
 * and produces the request SemanticGraph. Provider- and protocol-independent.
 */
public final class SemanticRequestResolver {
	private final SemanticContractSnapshot contract;

	public SemanticRequestResolver(SemanticContractSnapshot contract) {
		this.contract = Objects.requireNonNull(contract);
	}

	public SemanticGraph resolve(SemanticRequest request) {
		Objects.requireNonNull(request);
		var root = contract.get(request.root());
		if (request.selections().isEmpty())
			throw invalidSelection("A semantic request must contain at least one selection.");
		SemanticFilterValidator.validate(request.options() == null ? null : request.options().filter(), root, contract);
		var normalized = normalizeQueryOptions(request.options(), root);
		SemanticQueryOptionsValidator.validate(normalized, root);
		validateOrdering(normalized == null ? List.of() : normalized.effectiveOrder(), root, request.selections());

		var graph = new SemanticGraph();
		graph.setOptions(normalized);
		resolveSelections(root, request.selections(), graph, null, null, true);
		SemanticGraphValidator.validate(graph, contract);
		return graph;
	}

	private SemanticQueryOptions normalizeQueryOptions(SemanticQueryOptions options, SemanticEntity root) {
		if (options == null)
			return null;
		if (options.order() == null && options.after() == null)
			return options;
		var order = new ArrayList<SemanticOrderTerm>();
		for (var term : options.effectiveOrder())
			order.add(canonicalizeOrderTerm(term, root));
		if (options.after() != null) {
			var hasIdentity = order.stream().anyMatch(x -> x.path().isEmpty()
					&& x.aggregate() == SemanticOrderAggregate.NONE && x.field().equals(root.identity().fieldId()));
			if (!hasIdentity)
				order.add(new SemanticOrderTerm(root.identity().fieldId(), SemanticSortDirection.ASC));
		}
		return new SemanticQueryOptions(options.filter(), order, options.limit(), options.offset(), options.after());
	}

	private SemanticOrderTerm canonicalizeOrderTerm(SemanticOrderTerm term, SemanticEntity root) {
		if (term.aggregate() != SemanticOrderAggregate.COUNT || term.path().isEmpty())
			return term;
		var entity = root;
		for (var relationshipId : term.path()) {
			final var currentEntity = entity;
			var relationship = entity.relationships().stream().filter(x -> x.id().equals(relationshipId)).findFirst()
					.orElseThrow(() -> invalidSelection("Order relationship '" + relationshipId
							+ "' is not defined on '" + currentEntity.name() + "'."));
			entity = contract.get(relationship.target());
		}
		return new SemanticOrderTerm(entity.identity().fieldId(), term.direction(), term.path(), term.aggregate());
	}

	private void resolveSelections(SemanticEntity entity, List<SemanticSelection> selections, SemanticGraph graph,
			SemanticGraph.SemanticGraphNode parent, RelationshipId viaRelationship, boolean root) {
		var fields = new ArrayList<FieldId>();
		var relationships = new ArrayList<RelationshipSelection>();
		var relationshipIds = new HashSet<RelationshipId>();
		for (var selection : selections) {
			if (selection.field() != null && selection.relationship() != null)
				throw invalidSelection("A selection cannot contain both a field and a relationship.");
			if (selection.field() == null && selection.relationship() == null)
				throw invalidSelection("A selection must contain a field or a relationship.");
			if (selection.field() != null) {
				if (!selection.children().isEmpty())
					throw invalidSelection("Field '" + selection.field() + "' cannot have child selections.");
				if (!isDeclaredField(entity, selection.field()))
					throw invalidSelection(
							"Entity '" + entity.name() + "' does not declare field '" + selection.field() + "'.");
				if (!isFieldSelectable(entity, selection.field()))
					throw invalidSelection(
							"Field '" + entity.name() + "." + selection.field() + "' is not selectable.");
				if (!fields.contains(selection.field()))
					fields.add(selection.field());
			} else {
				var relationshipId = selection.relationship();
				if (!relationshipIds.add(relationshipId))
					throw invalidSelection("Relationship '" + relationshipId + "' is selected more than once on '"
							+ entity.name() + "'. Merge repeated selections in the adapter before resolution.");
				var relationship = entity.relationships().stream().filter(r -> r.id().equals(relationshipId))
						.findFirst().orElse(null);
				if (relationship == null)
					throw invalidSelection(
							"Entity '" + entity.name() + "' does not declare relationship '" + relationshipId + "'.");
				relationships.add(new RelationshipSelection(relationship, selection.children()));
			}
		}
		var node = root ? graph.addRoot(entity.id(), fields) : graph.add(entity.id(), viaRelationship, parent, fields);
		for (var selected : relationships) {
			var target = contract.get(selected.relationship().target());
			if (selected.children().isEmpty())
				throw invalidSelection("Relationship '" + entity.name() + "." + selected.relationship().name()
						+ "' requires child selections.");
			resolveSelections(target, selected.children(), graph, node, selected.relationship().id(), false);
		}
	}

	private void validateOrdering(List<SemanticOrderTerm> terms, SemanticEntity root,
			List<SemanticSelection> selections) {
		for (var term : terms) {
			var entity = root;
			var currentSelections = selections;
			SemanticRelationship finalRelationship = null;
			for (var relationshipId : term.path()) {
				final var currentEntity = entity;
				var relationship = entity.relationships().stream().filter(r -> r.id().equals(relationshipId))
						.findFirst().orElseThrow(() -> invalidSelection("Order relationship '" + relationshipId
								+ "' is not defined on '" + currentEntity.name() + "'."));
				if (relationship.cardinality() == RelationshipCardinality.MANY) {
					if (!relationshipId.equals(term.path().get(term.path().size() - 1)) || !term.isAggregate())
						throw new UnsupportedOperationException(
								"Ordering through collection relationship '" + entity.name() + "." + relationship.name()
										+ "' requires an explicit aggregate semantics (COUNT, MIN, or MAX).");
					if (term.aggregate() == SemanticOrderAggregate.MIN
							|| term.aggregate() == SemanticOrderAggregate.MAX) {
						var target = contract.get(relationship.target());
						if (!isDeclaredField(target, term.field()))
							throw invalidSelection("Aggregate order field '" + term.field() + "' is not defined on '"
									+ target.name() + "'.");
					}
				}
				finalRelationship = relationship;
				var aggregateCollectionHop = relationship.cardinality() == RelationshipCardinality.MANY
						&& term.isAggregate() && relationshipId.equals(term.path().get(term.path().size() - 1));
				if (!aggregateCollectionHop && !isRelationshipSelected(currentSelections, relationshipId))
					throw new IllegalStateException("Order path '"
							+ String.join(".", term.path().stream().map(Object::toString).toList())
							+ "' requires relationship '" + entity.name() + "." + relationship.name()
							+ "' to be selected. The provider will not introduce an implicit join solely for ordering.");
				entity = contract.get(relationship.target());
				currentSelections = findRelationshipSelections(currentSelections, relationshipId);
			}
			if (term.isAggregate()) {
				if (term.path().isEmpty())
					throw invalidSelection("Aggregate ordering requires a relationship path.");
				if (finalRelationship == null)
					throw invalidSelection("Aggregate order relationship '" + term.path().get(term.path().size() - 1)
							+ "' is not defined.");
				if (finalRelationship.cardinality() != RelationshipCardinality.MANY)
					throw invalidSelection("Aggregate ordering is only valid on collection relationships.");
				if ((term.aggregate() == SemanticOrderAggregate.MIN || term.aggregate() == SemanticOrderAggregate.MAX)
						&& !isFieldSortable(entity, term.field()))
					throw invalidSelection("Aggregate order field '" + term.field() + "' is not sortable.");
			} else if (!isDeclaredField(entity, term.field()))
				throw invalidSelection("Order field '" + term.field() + "' is not defined on '" + entity.name() + "'.");
			else if (!isFieldSortable(entity, term.field()))
				throw invalidSelection("Field '" + entity.name() + "." + term.field() + "' is not sortable.");
		}
	}

	private static boolean isFieldSelectable(SemanticEntity entity, FieldId fieldId) {
		if (entity.identity().fieldId().equals(fieldId))
			return true;
		return entity.fields().stream().anyMatch(
				x -> x.id().equals(fieldId) && (x.capabilities() & SemanticFieldCapabilities.SELECTABLE) != 0);
	}

	private static boolean isFieldSortable(SemanticEntity entity, FieldId fieldId) {
		if (entity.identity().fieldId().equals(fieldId))
			return true;
		return entity.fields().stream()
				.anyMatch(x -> x.id().equals(fieldId) && (x.capabilities() & SemanticFieldCapabilities.SORTABLE) != 0);
	}

	private static boolean isDeclaredField(SemanticEntity entity, FieldId fieldId) {
		return entity.identity().fieldId().equals(fieldId)
				|| entity.fields().stream().anyMatch(f -> f.id().equals(fieldId));
	}

	private static boolean isRelationshipSelected(List<SemanticSelection> selections, RelationshipId id) {
		return selections.stream().anyMatch(s -> id.equals(s.relationship()));
	}

	private static List<SemanticSelection> findRelationshipSelections(List<SemanticSelection> selections,
			RelationshipId id) {
		return selections.stream().filter(s -> id.equals(s.relationship())).findFirst().map(SemanticSelection::children)
				.orElse(List.of());
	}

	private static IllegalStateException invalidSelection(String message) {
		return new IllegalStateException("Invalid semantic request: " + message);
	}

	private record RelationshipSelection(SemanticRelationship relationship, List<SemanticSelection> children) {
	}
}

File: src/csharp/Foundgine.Core/Semantic/Resolution/SemanticRequestResolver.cs
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.Query;

namespace Foundgine.Core.Semantic.Resolution;

/// <summary>
/// Resolves a protocol-neutral SemanticRequest against the static semantic
/// model and produces the request SemanticGraph. This is deliberately
/// provider- and protocol-independent.
/// </summary>
public sealed class SemanticRequestResolver
{
    private readonly SemanticContractSnapshot _contract;

    public SemanticRequestResolver(SemanticContractSnapshot contract)
    {
        _contract = contract ?? throw new ArgumentNullException(nameof(contract));
    }

    public SemanticGraph Resolve(SemanticRequest request)
    {
        ArgumentNullException.ThrowIfNull(request);

        var root = _contract.Get(request.Root);

        if (request.Selections.Count == 0)
            throw InvalidSelection("A semantic request must contain at least one selection.");

        SemanticFilterValidator.Validate(request.Options?.Filter, root, _contract);
        var normalizedOptions = NormalizeQueryOptions(request.Options, root);
        SemanticQueryOptionsValidator.Validate(normalizedOptions, root);
        ValidateOrdering(normalizedOptions?.EffectiveOrder ?? [], root, request.Selections);

        var graph = new SemanticGraph { Options = normalizedOptions };

        ResolveSelections(root, request.Selections, graph, null, null, isRoot: true);
        SemanticGraphValidator.Validate(graph, _contract);

        return graph;
    }

    private SemanticQueryOptions? NormalizeQueryOptions(
        SemanticQueryOptions? options,
        SemanticEntity root)
    {
        if (options is null) return null;
        if (options.Order is null && options.After is null) return options;

        var order = options.EffectiveOrder
            .Select(term => CanonicalizeOrderTerm(term, root))
            .ToArray();

        // Cursor pagination requires a deterministic root ordering. The
        // identity field is the semantic tie-breaker and is added here,
        // rather than being invented by a provider later.
        if (options.After is not null)
        {
            var hasIdentityTieBreaker = order.Any(x =>
                x.EffectivePath.Count == 0 &&
                x.Aggregate == SemanticOrderAggregate.None &&
                x.Field == root.Identity.FieldId);

            if (!hasIdentityTieBreaker)
                order = order.Append(new SemanticOrderTerm(root.Identity.FieldId, SemanticSortDirection.Asc)).ToArray();
        }

        return options with { Order = order };
    }

    private SemanticOrderTerm CanonicalizeOrderTerm(SemanticOrderTerm term, SemanticEntity root)
    {
        if (term.Aggregate != SemanticOrderAggregate.Count || term.EffectivePath.Count == 0)
            return term;

        var entity = root;
        foreach (var relationshipId in term.EffectivePath)
        {
            var relationship = entity.Relationships.FirstOrDefault(x => x.Id == relationshipId)
                               ?? throw InvalidSelection(
                                   $"Order relationship '{relationshipId}' is not defined on '{entity.Name}'.");
            entity = _contract.Get(relationship.Target);
        }

        // Compatibility bridge: the current public order record retains a
        // FieldId for all aggregates, but COUNT has no target-field operand.
        // Canonicalize it to the target identity so providers never have to
        // interpret an arbitrary field attached to COUNT.
        return term with { Field = entity.Identity.FieldId };
    }

    private void ResolveSelections(
        SemanticEntity entity,
        IReadOnlyList<SemanticSelection> selections,
        SemanticGraph graph,
        SemanticGraphNode? parent,
        RelationshipId? viaRelationship,
        bool isRoot = false)
    {
        var fields = new List<FieldId>();
        var relationships = new List<(SemanticRelationship Relationship, IReadOnlyList<SemanticSelection> Children)>();
        var relationshipIds = new HashSet<RelationshipId>();

        foreach (var selection in selections)
        {
            if (selection.Field is not null && selection.Relationship is not null)
                throw InvalidSelection("A selection cannot contain both a field and a relationship.");

            if (selection.Field is null && selection.Relationship is null)
                throw InvalidSelection("A selection must contain a field or a relationship.");

            if (selection.Field is { } fieldId)
            {
                if (selection.Children.Count != 0)
                    throw InvalidSelection($"Field '{fieldId}' cannot have child selections.");

                if (!IsDeclaredField(entity, fieldId))
                    throw InvalidSelection($"Entity '{entity.Name}' does not declare field '{fieldId}'.");
                if (!IsFieldSelectable(entity, fieldId))
                    throw InvalidSelection($"Field '{entity.Name}.{fieldId}' is not selectable.");

                if (!fields.Contains(fieldId))
                    fields.Add(fieldId);

                continue;
            }

            var relationshipId = selection.Relationship!.Value;

            if (!relationshipIds.Add(relationshipId))
                throw InvalidSelection(
                    $"Relationship '{relationshipId}' is selected more than once on '{entity.Name}'. Merge repeated selections in the adapter before resolution.");

            var relationship = entity.Relationships.FirstOrDefault(r => r.Id == relationshipId);

            if (relationship is null)
                throw InvalidSelection(
                    $"Entity '{entity.Name}' does not declare relationship '{relationshipId}'.");

            relationships.Add((relationship, selection.Children));
        }

        var node = isRoot
            ? graph.AddRoot(entity.Id, fields)
            : graph.Add(entity.Id, viaRelationship, parent, fields);

        foreach (var (relationship, children) in relationships)
        {
            var target = _contract.Get(relationship.Target);

            if (children.Count == 0)
                throw InvalidSelection(
                    $"Relationship '{entity.Name}.{relationship.Name}' requires child selections.");

            ResolveSelections(
                target,
                children,
                graph,
                node,
                relationship.Id);
        }
    }

    private void ValidateOrdering(
        IReadOnlyList<SemanticOrderTerm> terms,
        SemanticEntity root,
        IReadOnlyList<SemanticSelection> selections)
    {
        foreach (var term in terms)
        {
            var entity = root;
            var currentSelections = selections;
            SemanticRelationship? finalRelationship = null;

            foreach (var relationshipId in term.EffectivePath)
            {
                var relationship = entity.Relationships.FirstOrDefault(r => r.Id == relationshipId)
                                   ?? throw InvalidSelection(
                                       $"Order relationship '{relationshipId}' is not defined on '{entity.Name}'.");

                if (relationship.Cardinality == RelationshipCardinality.Many)
                {
                    if (relationshipId != term.EffectivePath[^1] || !term.IsAggregate)
                    {
                        throw new NotSupportedException(
                            $"Ordering through collection relationship '{entity.Name}.{relationship.Name}' requires an explicit aggregate semantics (COUNT, MIN, or MAX).");
                    }

                    // COUNT describes the cardinality of the relationship; it
                    // does not semantically require a target field. Min/Max do.
                    if (term.Aggregate is SemanticOrderAggregate.Min or SemanticOrderAggregate.Max)
                    {
                        var target = _contract.Get(relationship.Target);
                        if (!IsDeclaredField(target, term.Field))
                            throw InvalidSelection(
                                $"Aggregate order field '{term.Field}' is not defined on '{target.Name}'.");
                    }
                }

                finalRelationship = relationship;

                var aggregateCollectionHop = relationship.Cardinality == RelationshipCardinality.Many &&
                                             term.IsAggregate &&
                                             relationshipId == term.EffectivePath[^1];

                if (!aggregateCollectionHop && !IsRelationshipSelected(currentSelections, relationshipId))
                {
                    throw new InvalidOperationException(
                        $"Order path '{string.Join(".", term.EffectivePath)}' requires relationship '{entity.Name}.{relationship.Name}' to be selected. " +
                        "The provider will not introduce an implicit join solely for ordering.");
                }

                entity = _contract.Get(relationship.Target);
                currentSelections = FindRelationshipSelections(currentSelections, relationshipId);
            }

            if (term.IsAggregate)
            {
                if (term.EffectivePath.Count == 0)
                    throw InvalidSelection("Aggregate ordering requires a relationship path.");

                if (finalRelationship is null)
                    throw InvalidSelection($"Aggregate order relationship '{term.EffectivePath[^1]}' is not defined.");

                if (finalRelationship.Cardinality != RelationshipCardinality.Many)
                    throw InvalidSelection("Aggregate ordering is only valid on collection relationships.");

                if (term.Aggregate is SemanticOrderAggregate.Min or SemanticOrderAggregate.Max &&
                    !IsFieldSortable(entity, term.Field))
                    throw InvalidSelection($"Aggregate order field '{term.Field}' is not sortable.");
            }
            else if (!IsDeclaredField(entity, term.Field))
            {
                throw InvalidSelection(
                    $"Order field '{term.Field}' is not defined on '{entity.Name}'.");
            }
            else if (!IsFieldSortable(entity, term.Field))
            {
                throw InvalidSelection($"Field '{entity.Name}.{term.Field}' is not sortable.");
            }
        }
    }

    private static bool IsFieldSelectable(SemanticEntity entity, FieldId fieldId) =>
        entity.Identity.FieldId == fieldId ||
        entity.Fields.FirstOrDefault(x => x.Id == fieldId)?.Capabilities
            .HasFlag(SemanticFieldCapabilities.Selectable) == true;

    private static bool IsFieldSortable(SemanticEntity entity, FieldId fieldId) =>
        entity.Identity.FieldId == fieldId ||
        entity.Fields.FirstOrDefault(x => x.Id == fieldId)?.Capabilities.HasFlag(SemanticFieldCapabilities.Sortable) ==
        true;

    private static bool IsDeclaredField(SemanticEntity entity, FieldId fieldId) =>
        entity.Identity.FieldId == fieldId || entity.Fields.Any(f => f.Id == fieldId);

    private static bool IsRelationshipSelected(
        IReadOnlyList<SemanticSelection> selections,
        RelationshipId relationshipId) =>
        selections.Any(s => s.Relationship == relationshipId);

    private static IReadOnlyList<SemanticSelection> FindRelationshipSelections(
        IReadOnlyList<SemanticSelection> selections,
        RelationshipId relationshipId)
    {
        var selection = selections.FirstOrDefault(s => s.Relationship == relationshipId);
        return selection?.Children ?? [];
    }

    private static InvalidOperationException InvalidSelection(string message) =>
        new($"Invalid semantic request: {message}");
}